### ISSUE TITLE
How to replace the docker-image resource when it is non-functioning?

### ISSUE THREAD

**Issue by KJTsanaktsidis** [concourse/atc#84]
I'm trying to install concourse using the binaries on ubuntu 16.04. It appears that on this version, both garden-linux and docker default to using overlayfs as their graph storage engine. Thus, the docker-image resource, running inside a garden container with an overlayfs root, tries to start the docker daemon with an overlayfs graph engine. This doesn't work - attempting to create a container errors with the following:

```
Step 1 : FROM docker.au.s522.net/centos-ruby
latest: Pulling from centos-ruby
(snip)
Digest: sha256:90e27d70f754556d6d2c0cc1f599a6ea7e0ae6a67d4e7abcb346bfa183ccb0a1
Status: Downloaded newer image for docker.au.s522.net/centos-ruby:latest
 ---> 42df1196e49a
Step 2 : WORKDIR /app
invalid argument
umount: can't unmount /var/lib/docker: Invalid argument
```

I guess this works on ubuntu 14.04 because it uses aufs instead of overlayfs. This is fine, because I don't really want to do this anyway - the performance of running two union filesystems inside each other is pretty horrific and makes my builds super slow.

What I *actually* want to do is mount a btrfs scratch volume inside the garden container, and start the docker daemon with that. It doesn't look like concourse exposes any way to bind directories from the worker host into the worker container (or does it? I'd love to be wrong), so you can't just bind-mount a btrfs directory via pipeline configuration.

Thankfully, garden-linux doesn't do a particularly great job at isolation, and seems to create containers without user namespace isolation, and with `CAP_SYS_ADMIN` enabled inside the container. I was able to hack up a version of the docker-image resource that basically did the following (in `common.sh`)

```
start_docker() {
  # <snip>
  # Mount the btrfs scratch volume from the host onto /var/lib/docker
  # This makes building containers f-a-s-t :)
  # First, break out of our jail by remounting devtmpfs and using that
  # to mount the underlying btrfs voluume
  mkdir /real_dev
  mkdir /scratch_volume
  mount -t devtmpfs devfmpfs /real_dev
  mount -t btrfs /real_dev/disk/by-label/scratch /scratch_volume
  # Make a unique name so we don't conflict with other dockers running
  # on the same host.
  # The mount namespace inode should do nicely :)
  NAMESPACE_INODE=$(readlink /proc/1/ns/mnt | sed 's/mnt://')
  mkdir "/scratch_volume/$NAMESPACE_INODE"
  # Bind it to /var/lib/docker
  mount --bind "/scratch_volume/$NAMESPACE_INODE" /var/lib/docker
  # and undo the damage
  umount /real_dev
  umount /scratch_volume

  server_args="${server_args} -s btrfs"
  docker daemon ${server_args} >/tmp/docker.log 2>&1 &
  # <snip>
}
```

This actually seemed to work (while I was hacking inside a `fly hijack` session). We remount the real /dev directory (which we can, because of CAP_SYS_ADMIN) and use it to get raw access to a block device to run the docker graph engine on. (By-the-by, I'd love to work out a way to expose the host docker socket inside the worker - it's not actually on the host root volume because `/var/run` is mounted as tmpfs. Any ideas?)

Unfortunately, I'm at my wits end on how to actually deploy this to my concourse server. All the documentation I've read deals with adding resource types using the docker-image resource, but of course, my docker-image resource is at the moment totally nonfunctional. I created a tarball with a `/rootfs` and `metadata.json` file and uploaded it to S3, thinking to do something like this:

```
resource_types:
  - name: docker-image
    type: s3
    source:
      access_key_id: <snip>
      secret_access_key: <snip>
      region_name: ap-southeast-2
      bucket: <snip>
      versioned_file: ci-docker-image-resource.tar
```

Whilst that *downloads* the tar file, it doesn't extract it to the output directory, so concourse just complains with "could not read file from tar" (presumably because concourse tarballed up the out directory of the s3 resource, which now contains a tarball, instead of a `rootfs` and `metadata.json`).

I'm completely out of ideas at this point. My next step is probably to start hacking at the worker code to try and use a resource off the host filesystem, but I'd really like to avoid that level of violence. I don't mind forking all the resources, but forking the ATC is probably a bit too far...

TL;DR: I have a tarball of the docker-image resource I want to use. How can I replace the existing docker-image resource, given that it is non-functional?

**Comment by vito** [comment]
You could have the worker advertise a new resource type, just as it advertises the core set. But at some level `resource_types` just can't work if there's no resource it can use to provide new ones.

I'm betting this is resolved by now though since we don't even use garden-linux or overlayfs anymore.
